lib/evmone/instructions.hpp

```diff
@@ -900,6 +900,29 @@ inline code_iterator swapn(StackTop stack, ExecutionState& state, code_iterator
     return pos + 2;
 }
 
+inline Result mcopy(StackTop stack, int64_t gas_left, ExecutionState& state) noexcept
+{
+    const auto& dst_index = stack.pop();
+    const auto& src_index = stack.pop();
+    const auto& size = stack.pop();
+
+    if (!check_memory(gas_left, state.memory, std::max(dst_index, src_index), size))
+        return {EVMC_OUT_OF_GAS, gas_left};
+
+    auto dst = static_cast<size_t>(dst_index);
+    auto src = static_cast<size_t>(src_index);
+    auto copy_size = static_cast<size_t>(size);
+
+    const auto copy_cost = num_words(copy_size) * 3;
+    if ((gas_left -= copy_cost) < 0)
+        return {EVMC_OUT_OF_GAS, gas_left};
+
+    if (copy_size > 0)
+        std::memcpy(&state.memory[dst], &state.memory[src], copy_size);
```

Comment:
From https://en.cppreference.com/w/cpp/string/byte/memcpy:
> If the objects overlap, the behavior is undefined.

Need to use something else.

Reply:
Yup, in C it does, I was wondering if in C++ it implies ownership changes. I guess not, it is just a counterpart of the C version.